### PR TITLE
interrupt_controller: gic: GIC Split Mode fix ups

### PR DESCRIPTION
**Pull request by bwasim**
During testing of R5 Split (performance) mode, the kernel was found non-functional due to the following problems in the GIC driver.

- All GIC interrupts were always routed to Core-0. This meant that any interrupts enabled by Core-1 (or others) are also routed to Core-0, and as Core-0 didn't have proper setup to support these interrupts, it ended up in a fault. This limitation is primarily because the Arm architecture hasn't really been tested with AMP support. 
- Multiples cores initializing the shared GIC controller meant that any initialization did by previous cores were over-ridden generating problems.

The first problem is fixed by reading the MPIDR register at runtime and getting the Core-ID from there. All this is done in the call to GIC IRQ_Enable routine, and so the IRQ will be enabled on the correct core always. Instead of overriding, we OR the cores, which means that the same interrupt can be routed to multiple CPUs if desired (SMP ?)

The second problem is fixed by adding a Kernel config option. This option will enable/disable dist init. The idea is that only a single image will enable this, whilst other will keep it disabled, and the
core which enables this option will act as the "master", and init the GIC distributor. This solution is only used for AMP systems. For SMP, as all cores share the same memory, we should use a global variable to ensure that only the "primary" core inits the distributor.

Signed-off-by: Bilal Wasim <bilalwasim676@gmail.com>

### CODE REVIEW

File: dts/arm/xilinx/zynqmp.dtsi [hunk 1]
```diff
@@ -30,6 +30,16 @@
...
+		uart1: uart@ff010000 {
+			compatible = "xlnx,xuartps";
+			reg = <0xff010000 0x4c>;
+			status = "disabled";
+			interrupts = <GIC_SPI 22 IRQ_TYPE_LEVEL
+					IRQ_DEFAULT_PRIORITY>;
+			interrupt-names = "irq_1";
+			label = "UART_1";
+		};
+
```
Comment (stephanosio) [inline dts/arm/xilinx/zynqmp.dtsi:42]: This should go into a separate commit.

**Review by stephanosio — CHANGES_REQUESTED** [review]
In general, I do not think the approach taken by this PR is something we want in the long run (to put it simply, it is too hacky).

What we would want is a proper OS-level framework to handle the AMP master-slave(s) topology for the SoCs that have shared peripherals across two or more CPUs; GIC here is just one particular instance of such and there may be more shared peripherals like L2/L3 cache controller, memory controller, etc.

The master processor in this case would be responsible for initialising all common/shared peripherals and sending IPI to the slave processor(s) to initiate the slave boot sequence (this should be implemented using the OpenAMP framework).

File: drivers/interrupt_controller/intc_gic.c [hunk 2]
```diff
@@ -14,13 +14,42 @@
...
 
+static inline unsigned int cpu_id(void)
+{
+	unsigned int mpidr = 0;
+
+#if (CONFIG_CPU_CORTEX_R == 1)
+	/* Read MPIDR and AND-out the cluster id etc. */
+	__asm volatile("mrc p15, 0, %0, c0, c0, 5" : "=r"(mpidr));
+	mpidr &= 0xF;
+#endif /* CONFIG_CPU_CORTEX_R == 1 */
```
[the remark this answers is not in the document]
Comment (bwasim) [reply drivers/interrupt_controller/intc_gic.c:25]: Yup, I agree that this needs to go into the arch specific files - May be a function named `arm_cpu_id_get` or `arch_cpu_id_get` or something ? The implementation I've done here is valid for v7-a / v7-r, and we would need to add similar code for v8-a, but that has its own semantics.. I will take care of this..

File: drivers/interrupt_controller/Kconfig.gic [hunk 3]
```diff
@@ -36,4 +36,12 @@ config GIC_VER
 	default 2 if GIC_V2
 	default 3 if GIC_V3
 
+config GIC_MASTER
+	bool "GIC Master core controls the GIC Distributor."
+	default y
+	help
+	  As GIC Distributor is shared among multiple cores, which may be running in AMP,
+	  it is necessary that only one of the cores configures/control GIC distributor.
+	  All other cores should only configure their core-specific GIC CPU interfaces.
```
[the remark this answers is not in the document]
Comment (bwasim) [reply drivers/interrupt_controller/Kconfig.gic:45]: So we want to move this option from GIC to the ZynqMP SoC Kconfig ? I'm fine with that, but as I mentioned before, I wouldn't want to tie the split mode support with OpenAMP framework, though that support would need to be added sometime soon as well..
